### DISCUSSION TITLE
Powerwall 2 Integration

### DISCUSSION THREAD

**Discussion by Nick-Sills** [springfall2008/batpred#1213] [General]
Hi,

@springfall2008 Thanks for creating this project. You seem to have gone 'all in' to solve the problems that I have already been thinking about, so it's much appreciated.

My setup is as follows:

2 x 3.24 kW PV arrays
SolarEdge Inverter
Tesla Powerwall 2
Tesla Gateway
Tesla Home Wall Charger - Car
Octopus Agile Import/Export Tarif

As far as I know, with all Powerwall installations, the Powerwall Gateway takes control of Battery charge and Discharge, as well monitoring Solar, House load, Grid load +/-,  Battery load +/-.   **The inverter is effectively a dumb input into the system, and is connected by a CT clamp to provide the AC Solar power monitoring** 

The Powerwall has 2 operational modes, `Self Powered`, and `Time Based`. The former is used in Summer months where solar exceeds or equals energy consumption, and the latter uses electricity rates to charge and discharge at the appropriate times, supposedly using Solar forecasting in its algorithm. `Time Based` isn't directly compatible Octopus Agile, but I've found a way to Charge and Discharge on demand, so I can set up HA Octopus Agile switch automations. 

I can flip the operational mode to `Time Based` and use my charge/discharge switches to force charge and discharge rather than rely on time slots.  

So far, so good, but the main problem with this is the lack of solar forecasting, and the balance between import and export rates, and house load to optimize when to charge and discharge.  This is what brought me here. 

The sensors and switches I have available (Some are from the Gateway, some are templates that I've created):

**Power**
House Load kW
Solar Power kW
Grid Import kW
Grid Export kW
Battery Power kW +/-

**Energy**
Daily House Load kWh
Daily Solar Power kWh
Daily Grid Import kWh
Daily Grid Export kWh
Daily Battery Charge kWh
Daily Battery Discharge kWh

**Battery**
SOC  %
SOC kWh

**Switches and Controls**
Battery Charge Switch  
Batter Discharge Switch
Car charge  Switch
Car charge power: Amps 0 - 32 variable

**Octopus**
1/2 hourly Import rates  £
1/2 hourly export rates  £

I have Solcast hobbyist account already configured .

There are sections in the YAML file that I can't seem to correlate, but I believe the above sensors and controls  should provide enough data and inputs to manually configure PredBat, is this correct? 

*I'm asking before trying, because I know what a rabbit hole it can be to see problems when you're trying something unfamiliar*

Thanks for any help.

**Comment by radjanoonan** [comment]
I have managed to populate a yaml file for powerwall and predbat using the sensors from homeassistant. I have gotten it to a state were predbat will run at least. Attached is the yaml I used.

No automations yet, and definitely no way to perform actions, but the dashboard and calculations seem to work so far in read only mode.
I have also set up a flat rate cost calculation. In your case, you should be able to set up an octopus rate as per the documentation examples on the predbat site. 

I am using both the Tesla integration and another one to connect to my solar inverter as well from homeassistant. If you have wifi on the solar edge inverter, you should be able to link Homeassistant to the solar inverter to get some read only sensors from it as well like I do in my case.
 
Apparently, there are ways to allow Homeassistant to control the Powerwall settings using the Tesla Fleet API, but I have not yet explored that option. Certainly, Tesla configuration is a bit limited, since it is not possible to change the rate of charge or set the maximum charge level like with other systems.

Replace the [Your_Powerwall_SiteName] with the name from your HomeAssistant sensors and copy the apps.yaml into your predbat installation. Also update to the relevant timezone and currency symbols. 

 I have made some comments in the relevant section in [ ] brackets.
[apps.1.yaml.txt](https://github.com/user-attachments/files/19925002/apps.1.yaml.txt)

**Comment by radjanoonan** [comment]
I found this blog, where someone has implemented a working setup using the Teslemitry API and Predbat. Notably, they have set up a custom inverter configuration specifically for Tesla.  you can check their confirm on their blog: https://edhull.co.uk/blog/2025-08-24/predbat-docker-tesla

**Reply by gcoan** [reply]
> I found this blog, where someone has implemented a working setup using the Tesla Fleet API and Predbat. Notably, they have set up a custom inverter configuration specifically for Tesla. you can check their confirm on their blog: https://edhull.co.uk/blog/2025-08-24/predbat-docker-tesla

The Ed Hull blog only covers charging the Powerwall, doesn't cover exporting from predbat at all.

An expanded version of the Powerwall configuration has been developed by @slee2112 see #2745 and an updated set of Predbat documentation has just been created

**Reply by Slee2112** [reply]
Just a note to say, this was developed with the powerwall 3 and Tesla Fleet API, i believe the default powerwall integration within home assistant is a local connection and some of the sensors might be different.

Happy to give you a hand to get it working though if it is different.
